standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1293,19 +1304,41 @@ public AllocateTableWriteIdsResponse allocateTableWriteIds(AllocateTableWriteIds
         lockInternal();
         dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
         stmt = dbConn.createStatement();
+        long writeId;
+        String s;
+
+        handle = getMutexAPI().acquireLock(MUTEX_KEY.WriteIdAllocator.name());
 
         if (rqst.isSetReplPolicy()) {
-          txnIds = getValidTxnIds(rqst, stmt, rs);
+          assert (rqst.isSetSrcTxnToWriteIdList());
+          assert (!rqst.isSetTxnIds());
+          txnIds = checkAndGetTargetTxnIds(rqst, stmt, rs);
+          writeId = rqst.getSrcTxnToWriteIdList().get(0).getWriteId();
         } else {
+          assert (!rqst.isSetSrcTxnToWriteIdList());
+          assert (rqst.isSetTxnIds());
           txnIds = rqst.getTxnIds();
+
+          // Get the next write id for the given table and update it with new next write id.
+          // This is select for update query which takes a lock if the table entry is already there in NEXT_WRITE_ID
+          s = sqlGenerator.addForUpdateClause(
+                  "select nwi_next from NEXT_WRITE_ID where nwi_database = " + quoteString(dbName)
+                          + " and nwi_table = " + quoteString(tblName));
+          LOG.debug("Going to execute query <" + s + ">");
+          rs = stmt.executeQuery(s);
+          if (!rs.next()) {
+            writeId = 1;
+          } else {
+            writeId = rs.getLong(1);;
+          }
         }
 
         if (txnIds.size() == 0) {
-          // Idempotent case for replication flow, return the same txn2writeId list as in request message.
-          return new AllocateTableWriteIdsResponse(rqst.getTxnToWriteIdList());
+          // Idempotent case for replication flow.
```

Comment:
Let's handle idempotent case for repl inside replPolicy code block.

Reply:
done